gcovr/configuration.py

```diff
@@ -39,32 +39,53 @@ def check_percentage(value):
     return x
 
 
-def check_input_file(value):
+def check_input_file(value, basedir=os.getcwd()):
     r"""
     Check that the input file is present. Return the full path.
     """
 
+    if not os.path.isabs(value):
+        value = os.path.join(basedir, value)
+    value = os.path.normpath(value)
+
     if not os.path.isfile(value):
         raise ArgumentTypeError(
             "Should be a file that already exists: {value!r}".format(value=value))
 
     return os.path.abspath(value)
 
 
-def check_output_file(value):
+def check_output(value, basedir=os.getcwd()):
     r"""
     Check if the output file can be created.
     """
 
     if value is not None:
-        try:
-            with open(value, 'w') as _:
-                pass
-            os.unlink(value)
-        except OSError as e:
-            raise ArgumentTypeError("Could not create output file {value!r}: {error}".format(value=value, error=e.strerror))
-
-        value = os.path.abspath(value)
+        value = str(value)
+        original_value = value
+        # Normalize the path. Don't use os.path.normpath because of removing the possible trailing slash
+        value = value.replace('\\', os.sep).replace('/', os.sep)
+        if value.endswith(os.sep):
+            try:
+                # Now mormalize and add the trailing slash after creating the directory.
+                if not os.path.isabs(value):
+                    value = os.path.join(basedir, value)
+                value = os.path.normpath(value)
+                if not os.path.isdir(value):
+                    os.mkdir(value)
+                value = os.path.normpath(value) + os.sep
+            except OSError as e:
+                raise ArgumentTypeError("Could not create output directory {value!r}: {error}".format(value=original_value, error=e.strerror))
+        else:
+            try:
+                if not os.path.isabs(value):
+                    value = os.path.join(basedir, value)
+                value = os.path.normpath(value)
+                with open(value, 'w') as _:
+                    pass
+                os.unlink(value)
+            except OSError as e:
+                raise ArgumentTypeError("Could not create output file {value!r}: {error}".format(value=original_value, error=e.strerror))
```

Comment:
>     * I think the condition can be replaced with `original_value.endswith('\\') or original_value.endswith('/')`, so that nothing has to be replaced. The `normpath()` is done later anyway.
OK

>     * Making the path absolute can also be moved out of the condition to avoid duplication:
>       ```python
>       if not os.path.isabs(value):
>           value = os.path.join(basedir, value)
>       value = os.path.normpath(value)
>       if ...:
>       ```
Then we have to add a flag if there was a trailing separator because making the path absolute removes the trailing separator.

>     * Could you explain why normpath() is more appropriate than abspath()?
abspath() doesn't accept a directory instead of using the working directory. Because of this, the join is done on our own and the path is absolute.

>     * Only the `os.mkdir()` and `with open()` statements need to be within the `try` block.
With the local flag about the trailing separator, this is true.

Reply:
@latk I've merged the `check_output_file` into the class `OutputOrDefault`. While testing the part with the configuration I found out, that the `isinstance` in the following two lines is not correct.
https://github.com/gcovr/gcovr/blob/1bc72e3bb59b9296e962b350691732ddafbd3195/gcovr/configuration.py#L338
https://github.com/gcovr/gcovr/blob/1bc72e3bb59b9296e962b350691732ddafbd3195/gcovr/configuration.py#L340
The reason is, that the first argument isn't an instance, it is a type itself and the result is always false.